File: leetcode/cpp/course-schedule-iv.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<bool> checkIfPrerequisite(int numCourses, vector<vector<int>>& prereq, vector<vector<int>>& queries) {
        vector<vector<int>> graph(numCourses);
        vector<int> indegree(numCourses,0);
        for(auto e : prereq){
            graph[e[0]].push_back(e[1]);
            indegree[e[1]];
        }
    }
};

File: leetcode/python/course-schedule-iv.py
from typing import List
from collections import defaultdict, deque
class Solution:
    def checkIfPrerequisite(self, numCourses: int, prerequisites: List[List[int]], queries: List[List[int]]) -> List[bool]:
        indegree = [0 for _ in range(numCourses)]
        adjGraph = defaultdict(set)
        for edge in prerequisites:
            u,v = edge[0], edge[1]
            indegree[v]+=1
            adjGraph[u].add(v)
        nodePrerequisites = defaultdict(set)
        q = deque()
        for i in range(numCourses):
            if indegree[i] == 0:
                q.append(i)
        while q:
            node = q.popleft()
            for adj in adjGraph[node]:
                nodePrerequisites[adj].add(node)
                for prereq in nodePrerequisites[node]:
                    nodePrerequisites[adj].add(prereq)
                indegree[adj]-=1
                if indegree[adj]==0:
                    q.append(adj)
        answer = []
        for query in queries:
            answer.append(query[0] in nodePrerequisites[query[1]])
        return answer